core/include/detray/tools/rk_stepper.hpp

```diff
@@ -0,0 +1,185 @@
+
+        // maximum trial number of RK stepping
+        size_t _max_rk_step_trials = 10000;
+
+        // accumulated path length
+        scalar _path_accumulated = 0;
+
+        // stepping data required for RKN4
+        struct stepping_data {
+            vector3 b_first, b_middle, b_last;
+            vector3 k1, k2, k3, k4;
+            array_t<scalar, 4> k_qop;
+        };
+
+        stepping_data _step_data;
+    };
+
+    /** Take a step, regulared by a constrained step
+     *
+     * @param s The state object that chaches
+     * @param es The external step, e.g. from navigation
+     *
+     * @return returning the heartbeat, indicating if the stepping is alive
+     */
+    DETRAY_HOST_DEVICE
+    bool step(state& stepping, scalar& es) {
+        auto& sd = stepping._step_data;
+
+        scalar error_estimate = 0;
+
+        // First Runge-Kutta point
+        sd.b_first =
+            _magnetic_field.get_field(stepping().pos(), context_type{});
+        sd.k1 = evaluatek(stepping, sd.b_first, 0);
+
+        const auto try_rk4 = [&](const scalar& h) -> bool {
+            // State the square and half of the step size
+            const scalar h2 = h * h;
+            const scalar half_h = h * 0.5;
+            auto pos = stepping().pos();
+            auto dir = stepping().dir();
+
+            // Second Runge-Kutta point
+            const vector3 pos1 = pos + half_h * dir + h2 * 0.125 * sd.k1;
+            sd.b_middle = _magnetic_field.get_field(pos1, context_type{});
+            sd.k2 = evaluatek(stepping, sd.b_middle, 1, half_h, sd.k1);
+
+            // Third Runge-Kutta point
+            sd.k3 = evaluatek(stepping, sd.b_middle, 2, half_h, sd.k2);
+
+            // Last Runge-Kutta point
+            const vector3 pos2 = pos + h * dir + h2 * 0.5 * sd.k3;
+            sd.b_last = _magnetic_field.get_field(pos2, context_type{});
+            sd.k4 = evaluatek(stepping, sd.b_last, 3, h, sd.k3);
+
+            // Compute and check the local integration error estimate
+            // @Todo
+            const auto err_vec = h2 * (sd.k1 - sd.k2 - sd.k3 + sd.k4);
+            error_estimate =
+                std::max(getter::norm(err_vec), static_cast<scalar>(1e-20));
+
+            return (error_estimate <= stepping._tolerance);
+        };
+
+        scalar step_size_scaling = 1.;
+        size_t n_step_trials = 0;
+
+        while (!try_rk4(stepping._step_size)) {
+            step_size_scaling = std::min(
+                std::max(0.25,
+                         std::sqrt(std::sqrt((stepping._tolerance /
+                                              std::abs(2. * error_estimate))))),
+                4.);
+
+            stepping._step_size = stepping._step_size * step_size_scaling;
+
+            // If step size becomes too small the particle remains at the
+            // initial place
+            if (stepping._step_size * stepping._step_size <
+                stepping._step_size_cutoff * stepping._step_size_cutoff) {
```

Comment:
Changed to `std::abs()`

Reply:
Ah, Ok, no that's not what I originally meant. I thought you needed the numeric value ```stepping._step_size_cutoff * stepping._step_size_cutoff```, which we could cache instead of recalculating over and over again. Calling ```std::abs``` might be even more costly than performing a multiplication, but it also makes the actual value of ```stepping._step_size_cutoff``` way more tangible. So lets leave it like this